src/stan/services/pathfinder/multi.hpp

```diff
@@ -105,136 +109,214 @@ inline int pathfinder_lbfgs_multi(
+  // Save idx of pathfinder and it's elbo for resampling later
+  tbb::concurrent_vector<std::pair<Eigen::Index, internal::elbo_est_t>>
+      elbo_estimates;
+  elbo_estimates.reserve(num_paths + 10);
+  auto constrain_fun = [](auto&& constrained_draws, auto&& unconstrained_draws,
+                          auto&& model, auto&& rng) {
+    model.write_array(rng, unconstrained_draws, constrained_draws);
+    return constrained_draws;
+  };
   try {
+    stan::callbacks::concurrent_writer safe_write{parameter_writer};
     tbb::parallel_for(
         tbb::blocked_range<int>(0, num_paths), [&](tbb::blocked_range<int> r) {
           for (int iter = r.begin(); iter < r.end(); ++iter) {
-            auto pathfinder_ret
-                = stan::services::pathfinder::pathfinder_lbfgs_single<true>(
-                    model, *(init[iter]), random_seed, stride_id + iter,
-                    init_radius, history_size, init_alpha, tol_obj, tol_rel_obj,
-                    tol_grad, tol_rel_grad, tol_param, num_iterations,
-                    num_elbo_draws, num_draws, save_iterations, refresh,
-                    interrupt, logger, init_writers[iter],
-                    single_path_parameter_writer[iter],
-                    single_path_diagnostic_writer[iter], calculate_lp);
-            if (unlikely(std::get<0>(pathfinder_ret) != error_codes::OK)) {
-              logger.error(std::string("Pathfinder iteration: ")
-                           + std::to_string(iter) + " failed.");
-              return;
+            if (psis_resample && calculate_lp) {
+              auto toss_write = [](auto&&... /* x */) {};
+              auto pathfinder_ret
+                  = stan::services::pathfinder::pathfinder_lbfgs_single<true>(
+                      model, *(init[iter]), random_seed, stride_id + iter,
+                      init_radius, history_size, init_alpha, tol_obj,
+                      tol_rel_obj, tol_grad, tol_rel_grad, tol_param,
+                      num_iterations, num_elbo_draws, num_draws,
+                      save_iterations, refresh, interrupt, logger,
+                      init_writers[iter], toss_write,
+                      single_path_diagnostic_writer[iter], calculate_lp,
+                      psis_resample);
+              if (unlikely(std::get<0>(pathfinder_ret) != error_codes::OK)) {
+                logger.error(std::string("Pathfinder iteration: ")
+                             + std::to_string(iter) + " failed.");
+                return;
+              }
+              lp_calls += std::get<2>(pathfinder_ret);
+              elbo_estimates.push_back(
+                  std::make_pair(iter, std::move(std::get<1>(pathfinder_ret))));
+            } else {
+              // For no psis, have single write to both single and multi writers
+              using multi_writer_t = stan::callbacks::multi_writer<
+                  SingleParamWriter,
+                  stan::callbacks::concurrent_writer<ParamWriter>>;
```

Comment:
Q: wouldn't the concurrent_writer being wrapped around the multi_writer be what we really want here?

Reply:
I think it can be either/or but I like flipping them so more work is put off on the other thread